### PR TITLE
Add srcset attribute for images

### LINKED ISSUE #3249: Use srcset for images to prevent loading full assets in editor and AMP stories
Unsplash has some huge images which users can then zoom into in their canvas, or resize to be small. In the output format (image/output.js) we should be at least using srcset to load the appropriate image sizes.

Testing instructions (copied from Pascal):
Given that the Unsplash integration is still behind a feature flag (and feature flags are not part of QA), focus by just adding some images from the library to the editor, saving & previewing the story, and verifying that you see srcset attributes for the images in the source code.

### PR DESCRIPTION
**Pull request by diegovar**
## Summary

This PR introduces the srcset attribute for all images (based on a resource) displayed in the editor.

## Relevant Technical Choices

The sizes attribute is omitted, because the size of image relative to the viewport cannot be determined. Given each of these images has an explicit size, the browser is able to obtain the right size image to display.

In the case of output.js, the amp-img tag uses the 'fill' layout, which with explicit width and height correctly autogenerates a valid 'sizes' attribute for the img tag.

## To-do

In the editor, an intermediate image size is used instead of the correct one according to the size of the image in the canvas. I believe this is because the editor adds the img tag and then resizes it at some point in the codebase.

This PR doesn't introduce the possibility of cropping the image when it's from Unsplash by using imgix params such as `rect`. I spent quite some time trying to get that to play nicely with how we display the image the AMP document and failed, but given that Unsplash are [not committed to supporting that param](https://unsplash.com/documentation#supported-parameters), I believe this improvement is enough for now.

## User-facing changes

No user visible changes should be seen, other than images (especially from Unsplash) displaying much faster and using much less bandwidth.

## Testing Instructions

Use the editor (especially by adding images from Unsplash via the `media3pTab` flag) and ensure that images display correctly in the editor and in the generated AMP stories.

---

Fixes #3249

### CODE REVIEW

**Comment by swissspidy** [comment]
Some tests would be nice here

File: assets/src/edit-story/elements/image/display.js [hunk 1]
```diff
@@ -53,6 +53,7 @@ function ImageDisplay({ element, box }) {
         ref={ref}
         draggable={false}
         src={resource.src}
+        srcSet={calculateSrcSet(resource)}
```
Comment (winterDroid) [inline assets/src/edit-story/elements/image/display.js:56]: Can resouce have an attribute `srcSet` to avoid creating it in multiple places?

Comment (diegovar) [reply assets/src/edit-story/elements/image/display.js:56]: Done.

Comment (diegovar) [reply assets/src/edit-story/elements/image/display.js:56]: I had to revert this change. Unfortunately doing it this was wouldn't be retroactive as the resource objects have already been created in the past.

**Comment by diegovar** [comment]
> Some tests would be nice here

Added a couple of tests for the files that already had tests.

**Review by winterDroid — APPROVED** [review]

**Comment by spacedmonkey** [comment]
Should we also apply source sets here https://github.com/google/web-stories-wp/blob/8ac0f3d8a7f3604fd751e73131a73f00a56dd678/assets/src/edit-story/components/library/panes/media/common/mediaElement.js#L357-L369

**Comment by diegovar** [comment]
> Should we also apply source sets here
> 
> https://github.com/google/web-stories-wp/blob/8ac0f3d8a7f3604fd751e73131a73f00a56dd678/assets/src/edit-story/components/library/panes/media/common/mediaElement.js#L357-L369

Not really necessary because the items shown in the media gallery are always at most 360px, and we call getThumbnailUrl passing in the actual width of the box where the media element will be displayed.

**Review by spacedmonkey — APPROVED** [review]
👍

**Review by swissspidy — CHANGES_REQUESTED** [review]
This can result in AMP invalidation.

Tested by uploading the following image:

architecture-blur-building-2225145

The resulting markup is as follows:

```html
<amp-img
layout="fill" 
width="330"
 height="220" 
 srcset="
 https://example.com/architecture-blur-building-2225145-scaled.jpg 2560w,
 https://example.com/architecture-blur-building-2225145-2048x1365.jpg 2048w,
 https://example.com/architecture-blur-building-2225145-1980x1320.jpg 1980w,
 https://example.com/architecture-blur-building-2225145-1536x1024.jpg 1536w,
 https://example.com/architecture-blur-building-2225145-1200x800.jpg 1200w,
 https://example.com/architecture-blur-building-2225145-1024x683.jpg 1024w,
 https://example.com/architecture-blur-building-2225145-928x928.jpg 928w,
 https://example.com/architecture-blur-building-2225145-928x696.jpg 928w,
 https://example.com/architecture-blur-building-2225145-768x512.jpg 768w,
 https://example.com/architecture-blur-building-2225145-696x928.jpg 696w,
 https://example.com/architecture-blur-building-2225145.jpg"
 alt="architecture-blur-building-2225145">
 </amp-img>
```

Note the two entries for 928w, caused by the various image sizes we have for the poster images (portrait / landscape / square ).

The implementation should remove duplicates. A test would be important for that scenario. Also,  a `.toBeValidAMPStoryElement()` assertion can't hurt

-----

Also:

When I dragged that image from my computer to the canvas, and later saved the story, no `srcset` attribute was added. Seems like we need to `updateElementsByResourceId()` after upload to ensure the dropped element gets the proper `srcset` attribute?

**Comment by diegovar** [comment]
> This can result in AMP invalidation.
> 
> Tested by uploading the following image:
> 
> architecture-blur-building-2225145
> 
> The resulting markup is as follows:
> 
> ```
> <amp-img
> layout="fill" 
> width="330"
>  height="220" 
>  srcset="
>  https://example.com/architecture-blur-building-2225145-scaled.jpg 2560w,
>  https://example.com/architecture-blur-building-2225145-2048x1365.jpg 2048w,
>  https://example.com/architecture-blur-building-2225145-1980x1320.jpg 1980w,
>  https://example.com/architecture-blur-building-2225145-1536x1024.jpg 1536w,
>  https://example.com/architecture-blur-building-2225145-1200x800.jpg 1200w,
>  https://example.com/architecture-blur-building-2225145-1024x683.jpg 1024w,
>  https://example.com/architecture-blur-building-2225145-928x928.jpg 928w,
>  https://example.com/architecture-blur-building-2225145-928x696.jpg 928w,
>  https://example.com/architecture-blur-building-2225145-768x512.jpg 768w,
>  https://example.com/architecture-blur-building-2225145-696x928.jpg 696w,
>  https://example.com/architecture-blur-building-2225145.jpg"
>  alt="architecture-blur-building-2225145">
>  </amp-img>
> ```
> 
> Note the two entries for 928w, caused by the various image sizes we have for the poster images (portrait / landscape / square ).
> 
> The implementation should remove duplicates. A test would be important for that scenario. Also, a `.toBeValidAMPStoryElement()` assertion can't hurt
> 
> Also:
> 
> When I dragged that image from my computer to the canvas, and later saved the story, no `srcset` attribute was added. Seems like we need to `updateElementsByResourceId()` after upload to ensure the dropped element gets the proper `srcset` attribute?

Thanks for finding that issue! I'll add a fix for it.

With regards to the last paragraph, the srcset is calculated and added in output.js (and not when creating the resource) so it shouldn't matter how the resource was added to the library. I guess what's happening is that the resource doesn't have a `sizes`?

**Comment by diegovar** [comment]
> updateElementsByResourceId

I've fixed the issue with duplicate widths in the AMP output and added tests.

With regards to the second issue, that seems unrelated to this (the fact that a dragged image doesn't generate sizes in getResourceFromLocalFile.js)

**Review by diegovar — COMMENTED** [review]
Done

**Review by swissspidy — CHANGES_REQUESTED** [review]
If there are no sizes, an empty `srcset=""` attribute will be output, which causes AMP invalidation.

It can be quickly tested by uploading the following small image:

tiny

It's an edge case, but we want to avoid AMP invalidation at all times.

Screenshot of the error:

Screenshot 2020-08-14 at 10 18 00

The img element on the frontend:

```html
<amp-img layout="fill" srcset="" src="https://stories-qa-wordpress-amp.pantheonsite.io/wp-content/uploads/2020/08/tiny.png" alt="tiny"></amp-img>
```

**Comment by swissspidy** [comment]
@diegovar Apologies for the back and forth on this one. I only spotted the issue when I saw the failed AMP validation test at https://github.com/google/web-stories-wp/runs/983576472.

Also, I created #3905 for the other issue I mentioned before about missing `srcset` when dragging an image from the computer onto the canvas.

**Comment by diegovar** [comment]
> @diegovar Apologies for the back and forth on this one. I only spotted the issue when I saw the failed AMP validation test at https://github.com/google/web-stories-wp/runs/983576472.
> 
> Also, I created #3905 for the other issue I mentioned before about missing `srcset` when dragging an image from the computer onto the canvas.

No need to apologise, thanks for finding it! I've fixed it and added a test for it.

**Review by swissspidy — APPROVED** [review]

**Pull request merged**